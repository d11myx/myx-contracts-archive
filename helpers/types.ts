import { IPool } from '../types';
import { eNetwork } from './constants';
import { IFundingRate } from '../types/contracts/core/FundingRate';
import type { PromiseOrValue } from '../types/common';
import type { BigNumberish } from 'ethers';

export interface SymbolMap<T> {
    [symbol: string]: T;
}

export type ParamsPerNetwork<T> = {
    [k in eNetwork]?: T;
};

export type TradingFeeTier = {
    takerFee: PromiseOrValue<BigNumberish>;
    makerFee: PromiseOrValue<BigNumberish>;
};

export interface PairInfoConfig {
    pairTokenDecimals: number;
    useWrappedNativeToken: boolean;
    pair: IPool.PairStruct;
    tradingConfig: IPool.TradingConfigStruct;
    tradingFeeConfig: IPool.TradingFeeConfigStruct & TradingFeeTier;
    fundingFeeConfig: IFundingRate.FundingFeeConfigStruct;
}

export interface ReserveConfiguration {
    MarketTokenSymbol: string;
    MarketTokenName: string;
    MarketTokenDecimals: number;
    MarketTokenAddress: ParamsPerNetwork<string>;
    WrapperTokenAddress: ParamsPerNetwork<string>;
    PairsConfig: SymbolMap<PairInfoConfig>;
    PairAssets: ParamsPerNetwork<SymbolMap<string>>;
    ExecuteOrderTimeDelay: number;
    OraclePriceFeedAddress: ParamsPerNetwork<string>;
    OraclePriceId: ParamsPerNetwork<SymbolMap<string>>;
    UniswapRouterAddress: ParamsPerNetwork<string>;
    UniswapTokenPathConfig: ParamsPerNetwork<SymbolMap<string>>;
}
