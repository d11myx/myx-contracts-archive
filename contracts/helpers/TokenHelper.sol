// SPDX-License-Identifier: BUSL-1.1
pragma solidity ^0.8.0;

import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
import "@openzeppelin/contracts/utils/math/SafeMath.sol";
import "../libraries/PrecisionUtils.sol";
import "../interfaces/IPool.sol";

library TokenHelper {
    using PrecisionUtils for uint256;
    using SafeMath for uint256;

    function convertIndexAmountToStable(
        IPool.Pair memory pair,
        int256 indexTokenAmount
    ) internal view returns (int256 amount) {
        uint256 indexTokenDec = uint256(IERC20Metadata(pair.indexToken).decimals());
        uint256 stableTokenDec = uint256(IERC20Metadata(pair.stableToken).decimals());

        uint256 indexTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - indexTokenDec);
        uint256 stableTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - stableTokenDec);

        amount = (indexTokenAmount * int256(indexTokenWad)) / int256(stableTokenWad);
    }

    // function convertIndexAmountToStableWithPrice(
    //     IPool.Pair memory pair,
    //     int256 indexTokenAmount,
    //     uint256 price
    // ) internal view returns (int256 amount) {
    //     uint256 indexTokenDec = uint256(IERC20Metadata(pair.indexToken).decimals());
    //     uint256 stableTokenDec = uint256(IERC20Metadata(pair.stableToken).decimals());

    //     uint256 indexTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - indexTokenDec);
    //     uint256 stableTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - stableTokenDec);

    //     amount =
    //         ((indexTokenAmount * int256(price) * int256(indexTokenWad)) /
    //             int256(PrecisionUtils.PRICE_PRECISION)) *
    //         int256(stableTokenWad);
    // }

    function convertStableAmountToIndex(
        IPool.Pair memory pair,
        int256 stableTokenAmount
    ) internal view returns (int256 amount) {
        uint256 indexTokenDec = uint256(IERC20Metadata(pair.indexToken).decimals());
        uint256 stableTokenDec = uint256(IERC20Metadata(pair.stableToken).decimals());

        uint256 indexTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - indexTokenDec);
        uint256 stableTokenWad = 10 ** (PrecisionUtils.maxTokenDecimals() - stableTokenDec);

        amount = (stableTokenAmount * int256(stableTokenWad)) / int256(indexTokenWad);
    }
}
