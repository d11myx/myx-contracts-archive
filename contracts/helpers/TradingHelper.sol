// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.0;

import '../libraries/PrecisionUtils.sol';
import '../interfaces/IAddressesProvider.sol';
import '../interfaces/IOraclePriceFeed.sol';
import '../interfaces/IIndexPriceFeed.sol';
import '../interfaces/IPool.sol';

// import 'hardhat/console.sol';

library TradingHelper {
    using PrecisionUtils for uint256;

    function getValidPrice(
        IAddressesProvider addressesProvider,
        address token,
        IPool.TradingConfig memory tradingConfig
    ) internal view returns (uint256) {
        uint256 oraclePrice = IOraclePriceFeed(addressesProvider.priceOracle()).getPrice(token);

        uint256 indexPrice = IIndexPriceFeed(addressesProvider.indexPriceOracle()).getPrice(token);

        uint256 diffP = oraclePrice > indexPrice ? oraclePrice - indexPrice : indexPrice - oraclePrice;
        diffP = diffP.calculatePercentage(oraclePrice);

        require(diffP <= tradingConfig.maxPriceDeviationP, 'exceed max price deviation');
        return oraclePrice;
    }
}
