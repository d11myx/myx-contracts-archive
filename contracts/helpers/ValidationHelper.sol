// SPDX-License-Identifier: BUSL-1.1
pragma solidity ^0.8.0;

import "../interfaces/IPositionManager.sol";
import "../interfaces/IPool.sol";
import "../interfaces/IAddressesProvider.sol";
import "../interfaces/IRoleManager.sol";
import "../libraries/PrecisionUtils.sol";
import "../libraries/TradingTypes.sol";

library ValidationHelper {
    using PrecisionUtils for uint256;

    function validateAccountBlacklist(
        IAddressesProvider addressesProvider,
        address account
    ) internal view {
        require(
            !IRoleManager(addressesProvider.roleManager()).isBlackList(account),
            "blacklist account"
        );
    }

    function validateOrderExpired(uint256 orderTime, uint256 maxTimeDelay) internal view {
        require(orderTime + maxTimeDelay >= block.timestamp, "order expired");
    }

    function validatePriceTriggered(
        IPool.TradingConfig memory tradingConfig,
        TradingTypes.TradeType tradeType,
        bool isAbove,
        uint256 currentPrice,
        uint256 orderPrice,
        uint256 maxSlippage
    ) internal pure {
        if (tradeType == TradingTypes.TradeType.MARKET) {
            bool valid = currentPrice >=
                orderPrice.mulPercentage(PrecisionUtils.percentage() - maxSlippage) &&
                currentPrice <= orderPrice.mulPercentage(PrecisionUtils.percentage() + maxSlippage);
            require(maxSlippage == 0 || valid, "exceeds max slippage");
        } else if (tradeType == TradingTypes.TradeType.LIMIT) {
            require(
                isAbove
                    ? currentPrice.mulPercentage(
                        PrecisionUtils.percentage() - tradingConfig.priceSlipP
                    ) <= orderPrice
                    : currentPrice.mulPercentage(
                        PrecisionUtils.percentage() + tradingConfig.priceSlipP
                    ) >= orderPrice,
                "not reach trigger price"
            );
        } else {
            require(
                isAbove ? currentPrice <= orderPrice : currentPrice >= orderPrice,
                "not reach trigger price"
            );
        }
    }
}
