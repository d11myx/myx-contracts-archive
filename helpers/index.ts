export * from './constants';
export * from './contract-deployments';
export * from './contract-getters';
export * from './deploy-ids';
export * from './env';
export * from './hardhat-config-helpers';
export * from './init-helper';
export * from './market-config-helper';
export * from './types';
export * from './utilities/signer';
export * from './utilities/tx';
export * from '../tasks/misc/print-deployments';

import { loadTasks } from './hardhat-config-helpers';

const TASK_FOLDERS = ['../tasks/misc'];

// Load all plugin tasks
loadTasks(TASK_FOLDERS);
