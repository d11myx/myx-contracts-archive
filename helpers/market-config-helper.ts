import { ReserveConfiguration } from './types';
import USDCMarketConfig from '../markets/usdc';

export enum ConfigNames {
    USDC = 'USDC',
}

export function loadReserveConfig(configName: ConfigNames): ReserveConfiguration {
    switch (configName) {
        case ConfigNames.USDC:
            return USDCMarketConfig;
        default:
            throw new Error(
                `Unsupported reserve configuration: ${configName} is not one of the supported configs ${Object.values(
                    ConfigNames,
                )}`,
            );
    }
}
