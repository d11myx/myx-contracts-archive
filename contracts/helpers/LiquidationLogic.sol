// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.20;

import "../libraries/Position.sol";
import "../interfaces/IPool.sol";
import "../interfaces/IOrderManager.sol";
import "../interfaces/IPositionManager.sol";
import "../interfaces/IAddressesProvider.sol";
import "../helpers/TradingHelper.sol";
import "../interfaces/IExecutor.sol";

library LiquidationLogic {
    using PrecisionUtils for uint256;
    using Position for Position.Info;

    uint256 public constant MAX_RATIO = 3e6;

    event ExecuteLiquidation(
        bytes32 positionKey,
        address account,
        uint256 pairIndex,
        bool isLong,
        uint256 collateral,
        uint256 sizeAmount,
        uint256 price
    );

    function liquidationPosition(
        IPool pool,
        IOrderManager orderManager,
        IPositionManager positionManager,
        IExecutor executor,
        IAddressesProvider addressesProvider,
        bytes32 positionKey,
        uint8 level,
        uint256 commissionRatio
    ) external {
        require(commissionRatio < MAX_RATIO, "max ratio");
        Position.Info memory position = positionManager.getPositionByKey(positionKey);
        if (position.positionAmount == 0) {
            return;
        }
        IPool.Pair memory pair = pool.getPair(position.pairIndex);
        IPool.TradingConfig memory tradingConfig = pool.getTradingConfig(position.pairIndex);

        uint256 price = TradingHelper.getValidPrice(
            addressesProvider,
            pair.indexToken,
            tradingConfig
        );

        int256 unrealizedPnl = position.getUnrealizedPnl(position.positionAmount, price);
        uint256 tradingFee = positionManager.getTradingFee(
            position.pairIndex,
            position.isLong,
            position.positionAmount
        );
        int256 fundingFee = positionManager.getFundingFee(
            position.account,
            position.pairIndex,
            position.isLong
        );
        int256 exposureAsset = int256(position.collateral) +
            unrealizedPnl -
            int256(tradingFee) +
            fundingFee;

        bool needLiquidate;
        if (exposureAsset <= 0) {
            needLiquidate = true;
        } else {
            uint256 riskRate = position
                .positionAmount
                .mulPrice(position.averagePrice)
                .mulPercentage(tradingConfig.maintainMarginRate)
                .calculatePercentage(uint256(exposureAsset));
            needLiquidate = riskRate >= PrecisionUtils.percentage();
        }
        if (!needLiquidate) {
            return;
        }

        // cancel all positionOrders
        orderManager.cancelAllPositionOrders(position.account, position.pairIndex, position.isLong);

        uint256 orderId = orderManager.createOrder(
            TradingTypes.CreateOrderRequest({
                account: position.account,
                pairIndex: position.pairIndex,
                tradeType: TradingTypes.TradeType.MARKET,
                collateral: 0,
                openPrice: price,
                isLong: position.isLong,
                sizeAmount: -int256(position.positionAmount),
                maxSlippage: 0,
                data: abi.encode(position.account)
            })
        );

        executor.executeDecreaseOrder(
            orderId,
            TradingTypes.TradeType.MARKET,
            level,
            commissionRatio
        );

        emit ExecuteLiquidation(
            positionKey,
            position.account,
            position.pairIndex,
            position.isLong,
            position.collateral,
            position.positionAmount,
            price
        );
    }
}
