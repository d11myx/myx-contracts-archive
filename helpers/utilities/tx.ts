import { BigNumber, Contract, ContractTransaction, ethers } from 'ethers';
import { HardhatRuntimeEnvironment } from 'hardhat/types';
import { TestEnv } from '../../test/helpers/make-suite';

const PRICE_PRECISION = '1000000000000000000000000000000';
const PERCENTAGE = '100000000';

declare var hre: HardhatRuntimeEnvironment;

export const waitForTx = async (tx: ContractTransaction) => await tx.wait(1);

export const deployContract = async <ContractType extends Contract>(
    contract: string,
    args?: any,
    libs?: { [libraryName: string]: string },
): Promise<ContractType> => {
    const [deployer] = await hre.ethers.getSigners();

    const contractFactory = await hre.ethers.getContractFactory(contract, {
        signer: deployer,
        libraries: {
            ...libs,
        },
    });

    const contractDeployed = await contractFactory.deploy(...args);

    return (await hre.ethers.getContractAt(contract, contractDeployed.address)) as any as ContractType;
};

// export const deployUpgradeableContract = async <ContractType extends Contract>(
//     contract: string,
//     args?: any,
// ): Promise<ContractType> => {
//     const [deployer] = await hre.ethers.getSigners();
//
//     const contractFactory = await hre.ethers.getContractFactory(contract, deployer);
//     let contractDeployed = await hre.upgrades.deployProxy(contractFactory, [...args]);
//
//     return (await hre.ethers.getContractAt(contract, contractDeployed.address)) as any as ContractType;
// };

export const getContract = async <ContractType extends Contract>(
    id: string,
    address?: string,
): Promise<ContractType> => {
    const artifact = await hre.deployments.getArtifact(id);
    return hre.ethers.getContractAt(
        artifact.abi,
        address || (await (await hre.deployments.get(id)).address),
    ) as any as ContractType;
};

interface AccountItem {
    name: string;
    account: string;
    balance: string;
}

export const getWalletBalances = async () => {
    const accounts = await hre.getNamedAccounts();

    const acc: AccountItem[] = [];
    for (let accKey of Object.keys(accounts)) {
        acc.push({
            name: accKey,
            account: accounts[accKey],
            balance: ethers.utils.formatEther(await hre.ethers.provider.getBalance(accounts[accKey])),
        });
    }
    return acc;
};

export const latestBlockNumber = async (): Promise<number> => {
    const block = await hre.ethers.provider.getBlock('latest');
    if (!block) {
        throw `latestBlockNumber: missing block`;
    }
    return block.number;
};

export const getBlockTimestamp = async (blockNumber?: number): Promise<number> => {
    if (!blockNumber) {
        const block = await hre.ethers.provider.getBlock('latest');
        if (!block) {
            throw `getBlockTimestamp: missing block number ${blockNumber}`;
        }
        return block.timestamp;
    }
    const block = await hre.ethers.provider.getBlock(blockNumber);
    if (!block) {
        throw `getBlockTimestamp: missing block number ${blockNumber}`;
    }
    return block.timestamp;
};

export async function latest() {
    const block = await hre.ethers.provider.getBlock('latest');
    return BigNumber.from(block.timestamp);
}

export async function latestBlock() {
    const block = await hre.ethers.provider.getBlock('latest');
    return BigNumber.from(block.number);
}

export async function advanceBlock() {
    await hre.ethers.provider.send('evm_mine', []);
}

export async function increase(duration: any) {
    if (!BigNumber.isBigNumber(duration)) {
        duration = BigNumber.from(duration);
    }

    // if (duration.isNeg()) throw Error(`Cannot increase time by a negative amount (${duration})`);

    await hre.ethers.provider.send('evm_increaseTime', [duration.toNumber()]);

    await advanceBlock();
}

export async function increaseTo(target: any) {
    if (!BigNumber.isBigNumber(target)) {
        target = BigNumber.from(target);
    }

    const now = await latest();

    if (target.lt(now)) throw Error(`Cannot increase current time (${now}) to a moment in the past (${target})`);
    const diff = target.sub(now);
    return increase(diff);
}

export const Duration = {
    seconds: function (val: any) {
        return BigNumber.from(val);
    },
    minutes: function (val: any) {
        return BigNumber.from(val).mul(this.seconds('60'));
    },
    hours: function (val: any) {
        return BigNumber.from(val).mul(this.minutes('60'));
    },
    days: function (val: any) {
        return BigNumber.from(val).mul(this.hours('24'));
    },
    weeks: function (val: any) {
        return BigNumber.from(val).mul(this.days('7'));
    },
    years: function (val: any) {
        return BigNumber.from(val).mul(this.days('365'));
    },
};

/**
 * calculation epoch funding rate
 *
 * @description (Clamp([w * (U - V) / K * Q + (1 - w) * (U - V) / K * L ] - Interest, min, max)/) / 365 / (24/4)
 * @param testEnv {TestEnv} current test env
 * @param pairIndex {Number} currency pair index
 * @returns funding rate
 */
export async function getFundingRate(testEnv: TestEnv, pairIndex: number) {
    const { positionManager, pool, oraclePriceFeed } = testEnv;
    const { indexTotalAmount, indexReservedAmount, stableTotalAmount, stableReservedAmount } = await pool.getVault(
        pairIndex,
    );

    const fundingInterval = 28800;
    const fundingFeeConfig = await pool.getFundingFeeConfig(pairIndex);
    const pair = await pool.getPair(pairIndex);
    const price = await oraclePriceFeed.getPrice(pair.indexToken);
    const exposedPosition = await positionManager.getExposedPositions(pairIndex);
    const longTracker = await positionManager.longTracker(pairIndex);
    const shortTracker = await positionManager.shortTracker(pairIndex);

    const uv = exposedPosition.abs().mul(price);
    const q = longTracker.add(shortTracker);
    const w = fundingFeeConfig.fundingWeightFactor;
    const k = fundingFeeConfig.liquidityPremiumFactor;
    const interest = fundingFeeConfig.interest;
    const diffBTCAmount = BigNumber.from(indexTotalAmount).sub(BigNumber.from(indexReservedAmount));
    const diffUSDTAmount = BigNumber.from(stableTotalAmount).sub(BigNumber.from(stableReservedAmount));
    const l = diffBTCAmount.mul(price).div(PRICE_PRECISION).add(diffUSDTAmount);

    let fundingRate, absFundingRate;
    if (q.eq(0)) {
        fundingRate = 0;
    } else {
        absFundingRate = BigNumber.from(w).mul(uv).mul(PERCENTAGE).div(k.mul(q));
        if (!l.eq(0)) {
            absFundingRate = absFundingRate.add(BigNumber.from(PERCENTAGE).sub(w).mul(uv).div(k.mul(l)));
        }
        fundingRate = exposedPosition.gte(0) ? absFundingRate : -absFundingRate;
    }

    fundingRate = BigNumber.from(fundingRate).sub(interest);
    fundingRate = fundingRate.lt(fundingFeeConfig.minFundingRate)
        ? fundingFeeConfig.minFundingRate
        : fundingRate.gt(fundingFeeConfig.maxFundingRate)
        ? fundingFeeConfig.maxFundingRate
        : fundingRate;

    return fundingRate.div(365).div(86400 / fundingInterval);
}

/**
 * calculation position average price
 *
 * @description averagePrice = (previousPositionAveragePrice * previousPositionAmount) + (openPrice * positionAmount) / (previousPositionAmount + positionAmount)
 * @param previousPositionAveragePrice {BigNumber} previous position average price
 * @param previousPositionAmount{BigNumber} previous position size
 * @param openPrice {BigNumber} current open position price
 * @param positionAmount {BigNumber} current position size
 * @returns position average price
 */
export function getAveragePrice(
    previousPositionAveragePrice: BigNumber,
    previousPositionAmount: BigNumber,
    openPrice: BigNumber,
    positionAmount: BigNumber,
) {
    return previousPositionAveragePrice
        .mul(previousPositionAmount)
        .add(openPrice.mul(positionAmount))
        .div(previousPositionAmount.add(positionAmount));
}

/**
 * calculation epoch funding fee tracker
 *
 * @description fundingFeeTracker = previousFundingFeeTracker * currentEpochFundingFee
 * @param previousFundingFeeTracker {BigNumber} previous epoch funding fee tracker
 * @param fundingRate {BigNumber} current epoch funding rate
 * @param openPrice {BigNumber} current epoch open price
 * @returns epoch funding fee tracker
 */
export function getFundingFeeTracker(
    previousFundingFeeTracker: BigNumber,
    fundingRate: BigNumber,
    openPrice: BigNumber,
) {
    return previousFundingFeeTracker.add(getEpochFundingFee(fundingRate, openPrice));
}

/**
 * calculation epoch currency standard funding fee
 *
 * @description currentEpochFundingFee = currentEpochRate * currentOpenPrice
 * @param fundingRate {BigNumber} current epoch funding rate
 * @param openPrice {BigNumber} current opsition open price
 * @returns each position holding one currency that passes the checkpoint requires a corresponding USDT fee to be paid
 */
export function getEpochFundingFee(fundingRate: BigNumber, openPrice: BigNumber) {
    return fundingRate.mul(openPrice).div(PRICE_PRECISION);
}

/**
 * calculation current position amount
 *
 * @param epochFundindFee {BigNumber} epoch currency standard funding fee
 * @param positionAmount {BigNumber} current position size
 * @returns current position amount
 */
export function getPositionFundFee(epochFundindFee: BigNumber, positionAmount: BigNumber) {
    return positionAmount.mul(epochFundindFee).div(PERCENTAGE);
}
