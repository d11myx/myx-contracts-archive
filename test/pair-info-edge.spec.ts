import { testEnv } from './helpers/make-suite';
import { describe } from 'mocha';
import { expect } from 'chai';
import { waitForTx } from './helpers/tx';
import { IPairInfo } from '../types/ethers-contracts';
import { loadPairConfig } from './helpers/market-config-helper';
import { BigNumber } from 'ethers';

describe('PairInfo: Edge cases', () => {
  before('addPair', async () => {
    const { pairInfo, btc, usdt } = testEnv;

    const btcPair = loadPairConfig('USDT', 'BTC');
    console.log('===========1');
    const pair = btcPair.pair;
    pair.indexToken = btc.address;
    pair.stableToken = usdt.address;
    const tradingConfig = btcPair.tradingConfig;
    const tradingFeeConfig = btcPair.tradingFeeConfig;
    const fundingFeeConfig = btcPair.fundingFeeConfig;

    console.log('===========2');

    await waitForTx(await pairInfo.addPair(pair, tradingConfig, tradingFeeConfig, fundingFeeConfig));

    console.log('===========3');
    console.log(await pairInfo.pairsCount());
    expect(await pairInfo.pairsCount()).to.be.eq('1');
    console.log('===========4');
  });

  it('check getters', async () => {
    console.log('11111111');
    const { pairInfo, pairLiquidity } = testEnv;
    console.log('2222222');

    expect(await pairInfo.pairLiquidity()).to.be.eq(pairLiquidity.address);
    console.log('3333333333');
  });

  describe('test addPair', async () => {
    it('check pair info', async () => {
      const { pairInfo, btc, usdt } = testEnv;

      const pairIndex = BigNumber.from(0);

      expect(await pairInfo.pairs(pairIndex)).deep.be.eq(await pairInfo.getPair(pairIndex));
      expect(await pairInfo.tradingConfigs(pairIndex)).deep.be.eq(await pairInfo.getTradingConfig(pairIndex));
      expect(await pairInfo.tradingFeeConfigs(pairIndex)).deep.be.eq(await pairInfo.getTradingFeeConfig(pairIndex));
      expect(await pairInfo.fundingFeeConfigs(pairIndex)).deep.be.eq(await pairInfo.getFundingFeeConfig(pairIndex));

      const btcPair = loadPairConfig('USDT', 'BTC');
      const pair = await pairInfo.getPair(pairIndex);
      expect(pair.indexToken).to.be.eq(btc.address);
      expect(pair.stableToken).to.be.eq(usdt.address);
      expect(pair.enable).to.be.eq(btcPair.pair.enable);
    });
  });

  describe('test updatePair', async () => {
    it('unHandler updatePair should be reverted', async () => {
      const {
        pairInfo,
        users: [unHandler],
      } = testEnv;

      const pairIndex = 0;
      const pair = await pairInfo.getPair(pairIndex);
      await expect(pairInfo.connect(unHandler.signer).updatePair(pairIndex, pair)).to.be.revertedWith(
        'Handleable: forbidden',
      );
    });

    it('check update pair', async () => {
      const { deployer, pairInfo } = testEnv;

      const pairIndex = 0;
      const pairBefore = await pairInfo.getPair(pairIndex);

      // updatePair
      let pairToUpdate: IPairInfo.PairStructOutput = { ...pairBefore };
      pairToUpdate.enable = !pairBefore.enable;
      pairToUpdate.kOfSwap = BigInt(99999999);
      pairToUpdate.initPairRatio = BigInt(999);
      await waitForTx(await pairInfo.connect(deployer.signer).updatePair(pairIndex, pairToUpdate));

      const pairAfterObj = await pairInfo.getPair(pairIndex);
      let pairAfter: IPairInfo.PairStructOutput = { ...pairAfterObj };

      console.log(pairAfter);
      console.log(pairToUpdate);
      // expect(pairAfter).deep.be.eq(pairToUpdate);

      // expect(pairAfter.enable).to.be.eq(pairToUpdate.enable);
      // expect(pairAfter.kOfSwap).to.be.eq(pairToUpdate.kOfSwap);
      // expect(pairAfter.initPairRatio).to.be.eq(pairToUpdate.initPairRatio);
    });
  });
});
