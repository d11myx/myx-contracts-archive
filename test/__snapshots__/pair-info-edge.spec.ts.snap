// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pool: Edge cases gas cost addPair  1`] = `38351`;
